File: sdk/src/privportal_sdk/identity.py
"""Cross-service identity resolution via PolarPrivate identity_bindings API.

Usage::

    from privportal_sdk import resolve_user, ServiceAuth

    # Without auth (public endpoints):
    result = resolve_user("feishu", "ou_abc123")

    # With bearer token auth:
    auth = ServiceAuth(token="ppst_...")
    result = resolve_user("feishu", "ou_abc123", auth=auth)
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

import httpx

if TYPE_CHECKING:
    from privportal_sdk.auth import ServiceAuth


def _default_base_url() -> str:
    port = os.environ.get("POLARPRIVATE_PORT", "12790")
    return f"http://127.0.0.1:{port}"


def _auth_headers(auth: ServiceAuth | None) -> dict[str, str]:
    if auth is None:
        return {}
    return auth.headers


@dataclass
class ResolvedUser:
    user_id: str
    username: str
    service: str
    external_username: str


def resolve_user(
    service: str,
    external_username: str,
    *,
    base_url: str | None = None,
    timeout: float = 5.0,
    auth: ServiceAuth | None = None,
) -> ResolvedUser | None:
    """Resolve an external service identity to a polarisor user_id.

    Returns None if no binding exists (404).
    Raises httpx.HTTPStatusError for other failures.
    """
    url = (base_url or _default_base_url()).rstrip("/")
    with httpx.Client(timeout=timeout, headers=_auth_headers(auth)) as client:
        resp = client.get(
            f"{url}/api/identity-bindings/resolve",
            params={"service": service, "external_username": external_username},
        )
        if resp.status_code == 404:
            return None
        resp.raise_for_status()
        data = resp.json()
        return ResolvedUser(**data)


async def aresolve_user(
    service: str,
    external_username: str,
    *,
    base_url: str | None = None,
    timeout: float = 5.0,
    auth: ServiceAuth | None = None,
) -> ResolvedUser | None:
    """Async variant of resolve_user."""
    url = (base_url or _default_base_url()).rstrip("/")
    async with httpx.AsyncClient(timeout=timeout, headers=_auth_headers(auth)) as client:
        resp = await client.get(
            f"{url}/api/identity-bindings/resolve",
            params={"service": service, "external_username": external_username},
        )
        if resp.status_code == 404:
            return None
        resp.raise_for_status()
        data = resp.json()
        return ResolvedUser(**data)


def list_user_bindings(
    user_id: str,
    *,
    base_url: str | None = None,
    timeout: float = 5.0,
    auth: ServiceAuth | None = None,
) -> list[dict[str, Any]]:
    """List all identity bindings for a given polarisor user_id."""
    url = (base_url or _default_base_url()).rstrip("/")
    with httpx.Client(timeout=timeout, headers=_auth_headers(auth)) as client:
        resp = client.get(f"{url}/api/identity-bindings/user/{user_id}")
        resp.raise_for_status()
        return resp.json().get("items", [])


def create_binding(
    user_id: str,
    service: str,
    external_username: str,
    *,
    display_name: str | None = None,
    base_url: str | None = None,
    timeout: float = 5.0,
    auth: ServiceAuth | None = None,
) -> dict[str, Any]:
    """Create a new identity binding."""
    url = (base_url or _default_base_url()).rstrip("/")
    body: dict[str, Any] = {
        "user_id": user_id,
        "service": service,
        "external_username": external_username,
    }
    if display_name:
        body["display_name"] = display_name
    with httpx.Client(timeout=timeout, headers=_auth_headers(auth)) as client:
        resp = client.post(f"{url}/api/identity-bindings", json=body)
        resp.raise_for_status()
        return resp.json()

File: sdk-ts/src/identity.ts
/**
 * Cross-service identity resolution via PolarPrivate identity_bindings API.
 *
 * @example
 * ```ts
 * import { resolveUser } from "@polarisor/privportal-sdk";
 * const user = await resolveUser("feishu", "ou_abc123");
 * // → { user_id: "...", username: "mac", service: "feishu", external_username: "ou_abc123" }
 * ```
 */

export interface ResolvedUser {
  user_id: string;
  username: string;
  service: string;
  external_username: string;
}

export interface IdentityBindingEntry {
  id: string;
  user_id: string;
  service: string;
  external_username: string;
  display_name: string | null;
  metadata_json: string | null;
  created_at: string;
  updated_at: string;
}

export interface IdentityBindingsListResponse {
  items: IdentityBindingEntry[];
  total: number;
}

function defaultBaseUrl(): string {
  const port = process.env.POLARPRIVATE_PORT ?? "12790";
  return `http://127.0.0.1:${port}`;
}

/**
 * Resolve an external service identity to a polarisor user_id.
 * Returns null if no binding exists.
 */
export async function resolveUser(
  service: string,
  externalUsername: string,
  baseUrl?: string,
): Promise<ResolvedUser | null> {
  const url = (baseUrl ?? defaultBaseUrl()).replace(/\/+$/, "");
  const params = new URLSearchParams({ service, external_username: externalUsername });
  const resp = await fetch(`${url}/api/identity-bindings/resolve?${params}`);
  if (resp.status === 404) return null;
  if (!resp.ok) throw new Error(`resolveUser failed: ${resp.status} ${await resp.text()}`);
  return resp.json();
}

/**
 * List all identity bindings for a given polarisor user_id.
 */
export async function listUserBindings(
  userId: string,
  baseUrl?: string,
): Promise<IdentityBindingEntry[]> {
  const url = (baseUrl ?? defaultBaseUrl()).replace(/\/+$/, "");
  const resp = await fetch(`${url}/api/identity-bindings/user/${userId}`);
  if (!resp.ok) throw new Error(`listUserBindings failed: ${resp.status} ${await resp.text()}`);
  const data: IdentityBindingsListResponse = await resp.json();
  return data.items;
}

/**
 * Create a new identity binding.
 */
export async function createBinding(
  userId: string,
  service: string,
  externalUsername: string,
  displayName?: string,
  baseUrl?: string,
): Promise<IdentityBindingEntry> {
  const url = (baseUrl ?? defaultBaseUrl()).replace(/\/+$/, "");
  const body: Record<string, string> = {
    user_id: userId,
    service,
    external_username: externalUsername,
  };
  if (displayName) body.display_name = displayName;
  const resp = await fetch(`${url}/api/identity-bindings`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  if (!resp.ok) throw new Error(`createBinding failed: ${resp.status} ${await resp.text()}`);
  return resp.json();
}
